unit tests for SelectionWriter file mode

tylerjereddy opened the pull request:
This PR adds a simple set of unit tests that ensure that `SelectionWriter` raises a `ValueError` when using an invalid file mode.

Generally speaking, a user is unlikely to encounter this particular error, but I decided to spend part of my Saturday evening looking through red lines in the coveralls report and found it [here](https://coveralls.io/builds/7924214/source?filename=miniconda%2Fenvs%2Fpyenv%2Flib%2Fpython2.7%2Fsite-packages%2FMDAnalysis%2Fselections%2Fbase.py#L104). So, a very minor increase in coverage, but ensures that this behaviour is retained moving forward.

richardjgowers commented:
Cool, I like the idea of microPRs to inch up coverage

orbeckst commented on package/CHANGELOG:
```diff
@@ -19,6 +19,7 @@ The rules for this file:
   * 0.16.0
 
 Enhancements
+  * Added unit tests for bad file mode passed to SelectionWriter
```
Admittedly, I am not looking at the full CHANGELOG at the moment and can't remember if we recorded tests in the past. However, I would say that pure tests should not be listed in CHANGELOG because it's not really very informative for most users.

By no means does this diminish the tons of kudos to you (and anyone) who takes the time to add tests!

You should still add your name to the list of contributors of this release, of course.

My $0.02.

mnmelo replied:
@jbarnoud and @fiona-naughton made extensive use of the testsuite CHANGELOG for this purpose.

orbeckst commented on testsuite/MDAnalysisTests/test_selections.py:
```diff
@@ -97,6 +97,10 @@ def test_writeselection_ndx(self):
         self._write_selection(name=self.ref_name)
         self._assert_indices()
 
+    def test_write_ndx_bad_mode(self):
```
Can you make this a generic method in `_SelectionWriter` instead of having almost the same method in each class?

orbeckst commented:
That's a good idea. 

--
Oliver Beckstein
email: orbeckst@gmail.com


> Am Sep 17, 2016 um 18:38 schrieb mnmelo <notifications@github.com>:
> 
> @jbarnoud and @fiona-naughton made extensive use of the testsuite CHANGELOG for this purpose

tylerjereddy commented:
Ok, I'll try to address some of the suggestions later today. Note that I looked into testing inheritance but the [discussions I found](http://stackoverflow.com/a/1323554) suggested that this actually introduces a fair bit of potential confusion / debate, and has drawbacks. So, fair warning that the beauty of abstracting the test may lead to different kinds of fragility, depending on how I do it, but it should probably be ok.

I'll think about moving the CHANGELOG testing information. I'm already a `0.16` author before my testing contributions because MDA couldn't properly write `.gro` files with more than 99,999 residues before the upcoming release.

tylerjereddy commented:
So, I've updated the PR in the following ways:

- The tests for the file mode have been condensed to a single test in the base class
- The package CHANGELOG note regarding this test (and a previous test I added for 0.16) has been moved to the test suite CHANGELOG, and my name has been added to the latter

Note that:

- I'm not clear on why the test suite CHANGELOG is for version `0.15.1` and the package CHANGELOG is currently targeting `0.16`, but I'm guessing this is just a bit of historical / current confusion about what the next release will be [please change that separately from this PR if you want to]
- I will generally assume that this strategy of using a private class in a testing module and inheriting from it is the standard for MDA, though I'm not convinced it is the most pythonic solution (that said, this is a matter of personal preference I think -- you can avoid running the base class itself by a number of other approaches, and if we use multiple inheritance then MRO matters for `setUp` and `tearDown`--yikes!)

richardjgowers commented:
It might be something we have to examine in the future, but I really like the pattern because it allows us to define the API as a class of tests. 

Thanks!
